### PR TITLE
feature: JSON Output

### PR DESCRIPTION
**Pull request by timkung**
# Provide JSON output for the content block

This adds functionality to output the data of the `ElementDecisionTree` as JSON to provide the ability to render custom front-end applications from the CMS configuration.

In essence, a developer will just need to call `->getBlockJSON()` on the instance of the content block and will receive the block data as JSON.

An example of the usage can be seen in the added `docs/en/json-output.md` file included in the PR.

### CODE REVIEW

File: src/Model/DecisionTreeAnswer.php [hunk 1]
```diff
@@ -181,4 +181,21 @@ public function getRecursiveEditPathForSelf()
...
+     * @return array
+     */
+    public function toJSONData()
+    {
+        $data = [
+            'title' => $this->Title,
+            'goTo' => $this->ResultingStepID,
+        ];
+
+        $this->extend('updateJSONData', $data);
```
Comment (jakxnz) [inline src/Model/DecisionTreeAnswer.php:197]: 🙏🏻

File: src/Model/ElementDecisionTree.php [hunk 2]
```diff
@@ -75,4 +76,89 @@ public function CMSEditFirstStepLink()
...
+
+        if (!$first->exists()) {
+            return $data;
+        }
+
+        // now start our descent through the flow
+        $this->collectStepData($first, $data);
+
+        // provide way to inject extra data
+        $this->extend('updateTreeData', $data);
```
Comment (jakxnz) [inline src/Model/ElementDecisionTree.php:120]: 🙏🏻

File: src/Model/ElementDecisionTree.php [hunk 3]
```diff
@@ -75,4 +76,89 @@ public function CMSEditFirstStepLink()
...
+     * Helper to collect the data for a given step.
+     *
+     * @param DecisionTreeStep $step
+     * @param array &$data
+     * @return void
+     */
+    private function collectStepData($step, &$data)
+    {
+        // add step data into array
+        $data['steps'][$step->ID] = $step->toJSONData();
```
Comment (jakxnz) [inline src/Model/ElementDecisionTree.php:135]: I'm aware this suggestion might mess up any existing implementations of this solution; but given that this is intended for JS, is it worthwhile not using an index here, and instead allow JS to parse this into an array? An object of index properties feels like a less predictable js object structure.

File: src/Model/ElementDecisionTree.php [hunk 4]
```diff
@@ -75,4 +76,89 @@ public function CMSEditFirstStepLink()
...
+     * @param DataList $answers
+     * @param array &$data
+     * @return void
+     */
+    private function collectAnswersData($answers, &$data)
+    {
+        // loop through answers
+        foreach ($answers as $answer) {
+            // push answer data into array
+            $data['answers'][$answer->ID] = $answer->toJSONData();
```
Comment (jakxnz) [inline src/Model/ElementDecisionTree.php:153]: Same question here, re not using an index here?

File: src/Model/ElementDecisionTree.php (hunk 3 above)
Comment (timkung) [reply src/Model/ElementDecisionTree.php:135]: Found this was the better approach as it is faster and easier when it comes to finding the data for the next step/answer to render, e.g. you loop to find the first step to render, that gives you `answers` as an array of ids for the answers data, so you can then just extract out the answer data referencing those ids as keys from the `answers` data vs it being an array and having to loop the whole answers array multiple times to find the data you are looking for.

File: src/Model/ElementDecisionTree.php (hunk 4 above)
Comment (timkung) [reply src/Model/ElementDecisionTree.php:153]: as above. and the answer data contains a `goTo`, which is the id of the next step, so you can just reference it directly by key in the `steps` data.

**Comment by scott1702** [comment]
Hey @alex-dna just checking if this is something you'd want as an option in the core module? We may look at splitting this logic into a separate module if it's not something you're keen on.

**Comment by alex-dna** [comment]
@scott1702 @timkung Thank you for your contribution. This sounds like a good idea (I've been willing to remove any dependency to js libraries, and having a plain json output to work from is the way to go). Unfortunately this is not up to me to merge this feature. I'm trying to get someone to approve or decline as this module is used on quite a few of our sites. Also, we would perhaps consider an approach using GraphQL. 
So I don't think this is going to be merged into the core anytime soon. If you need this feature, perhaps it could be implemented as an extension.

**Comment by scott1702** [comment]
Just wanting to update that we've built this extension out as a separate module if anyone wants to use this instead of the fork: https://github.com/silverstripeltd/silverstripe-elemental-decisiontree-json. Will leave this PR open for the owner of this module to either merge or close depending on whether they want this as part of the module 🤙

**Pull request closed without merging**